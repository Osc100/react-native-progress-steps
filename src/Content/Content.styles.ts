import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  container: {
    paddingBottom: 20,
    paddingLeft: 5,
    overflow: 'hidden',
  },
  content: {
    position: 'absolute',
  },
});

export default styles;
