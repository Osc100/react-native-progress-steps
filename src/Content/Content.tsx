import { Animated, LayoutChangeEvent, View } from 'react-native';
import React, { FC, memo, useEffect, useRef, useState } from 'react';

import { EASING } from '../constants';
import type { IContent } from './Content.types';
import useStyles from './Content.styles';

const Content: FC<IContent> = ({
  children,
  orientation,
  stepState: { isActive, isFirstInteraction } = {},
}) => {
  const [contentHeight, setContentHeight] = useState(0);
  const heightAnimation = useRef(new Animated.Value(0)).current;
  const opacityAnimation = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    Animated.parallel(
      [
        Animated.timing(heightAnimation, {
          toValue: isActive ? 1 : 0,
          duration: 400,
          easing: EASING,
          useNativeDriver: false,
        }),
        Animated.timing(opacityAnimation, {
          toValue: isActive ? 1 : 0,
          duration: isActive ? 400 : 200,
          delay: isActive ? 50 : 0,
          easing: EASING,
          useNativeDriver: false,
        }),
      ],
      {
        stopTogether: false,
      }
    ).start();
  }, [isActive]); // eslint-disable-line react-hooks/exhaustive-deps

  const handleOnLayout = (el: LayoutChangeEvent) => {
    const { height } = el.nativeEvent.layout;

    setContentHeight(height);
  };

  const opacity = isFirstInteraction && isActive ? 1 : opacityAnimation;
  const height =
    isFirstInteraction && isActive
      ? contentHeight
      : heightAnimation.interpolate({
          inputRange: [0, 1],
          outputRange: [0, contentHeight],
        });
  const styles = useStyles(orientation);
  return (
    <Animated.View
      pointerEvents={isActive ? 'auto' : 'none'}
      style={[styles.container, { opacity, height }]}
    >
      <View style={styles.content} onLayout={handleOnLayout}>
        {children}
      </View>
    </Animated.View>
  );
};

export default memo(
  Content,
  (prevProps, nextProps) =>
    prevProps.stepState?.isActive === nextProps.stepState?.isActive
);
