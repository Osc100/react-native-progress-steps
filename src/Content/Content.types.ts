import type { IstepState } from '../ProgressSteps.types';
import type React from 'react';

export interface IContent {
  children: React.ReactElement | React.ReactElement[];
  stepState?: IstepState;
}
