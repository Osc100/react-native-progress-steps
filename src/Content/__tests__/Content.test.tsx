import Content from '..';
import React from 'react';
import { Text } from 'react-native';
import { render } from '@testing-library/react-native';

const stepState = {
  stepIndex: 0,
  isActive: true,
  isCompleted: false,
  isFirst: true,
  isLast: false,
  isFirstInteraction: true,
};

describe('Title', () => {
  it('should render Title', () => {
    const { queryByText, toJSON } = render(
      <Content stepState={stepState}>
        <Text>Hello content</Text>
      </Content>
    );

    expect(queryByText('Hello content')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });
});
