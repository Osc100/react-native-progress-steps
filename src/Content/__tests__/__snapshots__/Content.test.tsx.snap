// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title should render Title 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "height": 0,
      "opacity": 1,
      "overflow": "hidden",
      "paddingBottom": 20,
      "paddingLeft": 5,
      "width": "100%",
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Text>
      Hello content
    </Text>
  </View>
</View>
`;
